victor/src/layout/fragments.rs

```diff
@@ -53,3 +64,39 @@ impl BoxFragment {
             .inflate(&self.border)
     }
 }
+
+impl CollapsingContext {
+    pub fn from_margin(margin: &Sides<Length>) -> Self {
+        Self {
+            start: CollapsedMargin::new(margin.block_start),
+            end: CollapsedMargin::new(margin.block_end),
+        }
+    }
+}
+
+impl CollapsedMargin {
+    pub fn zero() -> Self {
+        Self {
+            max_positive: Length::zero(),
+            min_negative: Length::zero(),
+        }
+    }
+
+    pub fn new(margin: Length) -> Self {
+        Self {
+            max_positive: margin.max(Length::zero()),
+            min_negative: margin.min(Length::zero()),
+        }
+    }
+
+    pub fn adjoin(&self, other: &Self) -> Self {
+        Self {
+            max_positive: self.max_positive.max(other.max_positive),
+            min_negative: self.min_negative.min(other.min_negative),
+        }
+    }
+
+    pub fn solve(&self) -> Length {
+        self.max_positive - self.min_negative
```

Comment:
Should this be a `+`, if `min_negative` is indeed negative and not the absolute value?

Reply:
Yes, will fix that.